jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,116 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.ByteBuffer;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public interface Index<V> extends OpaqueIndex
+{
+    @Override
+    default boolean contains(String key)
+    {
+        return get(key) != null;
+    }
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @return the value for the string key
+     */
+    V get(String s);
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @param offset The offset within the string of the key
+     * @param len the length of the key
+     * @return the value for the string / offset / length
+     */
+    V get(String s, int offset, int len);
+
+    /**
+     * Get the best match from key in a byte buffer.
+     * The key is assumed to by ISO_8859_1 characters.
+     *
+     * @param b The buffer
+     * @param offset The offset within the buffer of the key
+     * @param len the length of the key
+     * @return The value or null if not found
+     */
+    V getBest(ByteBuffer b, int offset, int len);
+
+    class Builder<V>
+    {
+        private Map<String, V> contents = new LinkedHashMap<>();
+        private boolean caseSensitive;
+
+        public Builder<V> caseSensitive(boolean caseSensitive)
+        {
+            this.caseSensitive = caseSensitive;
+            return this;
+        }
+
+        public <T> Builder<V> withAllDerivate(T[] values, Function<T, Map<String, V>> function)
```

Comment:
wow that's a name!!!  why not just overload `withAll`?

Reply:
This method is certainly going to be replaced with some form of `withAll()`, I don't like it either.